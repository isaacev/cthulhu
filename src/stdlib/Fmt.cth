#[inline]
fn int(i: Int) -> Str {
  ::Kernel::Builtins::cast_int_to_string(i)
}

#[inline]
fn float(f: Float) -> Str {
  ::Kernel::Builtins::cast_float_to_string(f)
}
