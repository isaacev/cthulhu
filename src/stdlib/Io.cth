intrinsic {
  fn php_print(Str) -> ();
  fn read_argv() -> [Str];
}

#[inline]
pub fn print(str: Str) -> () {
  php_print(str)
}

#[inline]
pub fn println(str: Str) -> () {
  php_print(str ++ "\n")
}

pub fn argv() -> [Str] {
  read_argv()
}
