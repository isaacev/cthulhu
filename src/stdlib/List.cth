use ::Kernel::Types::*;

#[inline]
fn nth(ls: ['a], n: Int) -> 'a {
  ::Kernel::Builtins::subscript(ls, n)
}
