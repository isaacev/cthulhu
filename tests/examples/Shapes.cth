use ::Io;
use ::Fmt;
use ::List;

type Shape =
  | UnitCircle
  | Circle(Float)
  | Square(Float)
  | Rect { width: Float, height: Float };

fn name(sh: Shape) -> Str {
  match sh {
    UnitCircle => "unit circle",
    Circle(1.0) => "unit circle",
    Circle(_) => "circle",
    Square(_) => "square",
    Rect { width: _, height: _ } => "rectangle",
  }
}

fn area(sh: Shape) -> Float {
  match sh {
    UnitCircle => 3.14,
    Circle(1.0) => 3.14,
    Circle(r) => 3.14 *. r *. r,
    Square(s) => s *. s,
    Rect { width: w, height: h } => w *. h,
  }
}

fn desc(sh: Shape) -> () {
  Io::println(name(sh) ++ " has an area of " ++ Fmt::float(area(sh)));
}

#[entry]
fn main() -> () {
  let u = UnitCircle;
  let v = Circle(1.0);
  let c = Circle(5.0);
  let s = Square(3.0);
  let r = Rect { width: 4.0, height: 3.0 };

  [ u, v, c, s, r] |> List::iter(desc);
}
