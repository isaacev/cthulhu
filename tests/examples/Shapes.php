<?php

namespace _List {
  function iter($b, $c) {
    while (true) {
      $callback = $b;
      $inputs = $c;
      if (\count($inputs) == 0) {
        return null;
      } else if (\count($inputs) >= 1) {
        $first = $inputs[0];
        $rest = \array_slice($inputs, 1);
        $callback($first);
        $b = $callback;
        $c = $rest;
        continue;
      } else {
        die("match expression did not cover all possibilities\n");
      }
    }
  }
}

namespace Shapes {
  abstract class Shape {}
  class UnitCircle extends \Shapes\Shape {
    function __construct() {
      // empty
    }
  }
  class Circle extends \Shapes\Shape {
    function __construct($a) {
      $this->{0} = $a;
    }
  }
  class Square extends \Shapes\Shape {
    function __construct($a) {
      $this->{0} = $a;
    }
  }
  class Rect extends \Shapes\Shape {
    public $width;
    public $height;
    function __construct($a) {
      $this->width = $a["width"];
      $this->height = $a["height"];
    }
  }
  function desc($sh) {
    if ($sh instanceof \Shapes\UnitCircle) {
      $a = "unit circle";
    } else if ($sh instanceof \Shapes\Circle && $sh->{0} == 1.0) {
      $a = "unit circle";
    } else if ($sh instanceof \Shapes\Circle) {
      $a = "circle";
    } else if ($sh instanceof \Shapes\Square) {
      $a = "square";
    } else if ($sh instanceof \Shapes\Rect) {
      $a = "rectangle";
    } else {
      die("match expression did not cover all possibilities\n");
    }
    if ($sh instanceof \Shapes\UnitCircle) {
      $c = 3.14;
    } else if ($sh instanceof \Shapes\Circle && $sh->{0} == 1.0) {
      $c = 3.14;
    } else if ($sh instanceof \Shapes\Circle) {
      $r = $sh->{0};
      $c = 3.14 * $r * $r;
    } else if ($sh instanceof \Shapes\Square) {
      $s = $sh->{0};
      $c = $s * $s;
    } else if ($sh instanceof \Shapes\Rect) {
      $w = $sh->width;
      $h = $sh->height;
      $c = $w * $h;
    } else {
      die("match expression did not cover all possibilities\n");
    }
    print($a . " has an area of " . (string)$c . "\n");
    return null;
  }

  function main() {
    $u = new \Shapes\UnitCircle();
    $v = new \Shapes\Circle(1.0);
    $c = new \Shapes\Circle(5.0);
    $s = new \Shapes\Square(3.0);
    $r = new \Shapes\Rect([
      "width" => 4.0,
      "height" => 3.0
    ]);
    \_List\iter('\Shapes\desc', [
      $u,
      $v,
      $c,
      $s,
      $r
    ]);
    return null;
  }
}

namespace {
  \Shapes\main(null);
}
