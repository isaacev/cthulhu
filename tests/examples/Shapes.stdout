unit circle has an area of 3.14
unit circle has an area of 3.14
circle has an area of 78.5
square has an area of 9
rectangle has an area of 12
